test/hotspot/jtreg/compiler/lib/template_framework/Renderer.java

```diff
@@ -0,0 +1,398 @@
+                // definitions at the hook cannot escape the hookCodeFrame.
+                CodeFrame hookCodeFrame = CodeFrame.make(outerCodeFrame);
+                hookCodeFrame.addHook(hook);
+
+                // We need a CodeFrame where the tokens can be rendered. That way, name
+                // definitions from the tokens cannot escape the innerCodeFrame to the
+                // hookCodeFrame.
+                CodeFrame innerCodeFrame = CodeFrame.make(hookCodeFrame);
+                currentCodeFrame = innerCodeFrame;
+
+                renderTokenList(tokens);
+
+                // Close the hookCodeFrame and innerCodeFrame. hookCodeFrame code comes before the
+                // innerCodeFrame code from the tokens.
+                currentCodeFrame = outerCodeFrame;
+                currentCodeFrame.addCode(hookCodeFrame.getCode());
+                currentCodeFrame.addCode(innerCodeFrame.getCode());
+            }
+            case HookInsertToken(Hook hook, TemplateToken t) -> {
+                // Switch to hook CodeFrame.
+                CodeFrame callerCodeFrame = currentCodeFrame;
+                CodeFrame hookCodeFrame = codeFrameForHook(hook);
+
+                // Use a transparent nested CodeFrame. We need a CodeFrame so that the code generated
+                // by the TemplateToken can be collected, and hook insertions from it can still
+                // be made to the hookCodeFrame before the code from the TemplateToken is added to
+                // the hookCodeFrame.
+                // But the CodeFrame must be transparent, so that its name definitions go out to
+                // the hookCodeFrame, and are not limited to the CodeFrame for the TemplateToken.
+                currentCodeFrame = CodeFrame.makeTransparentForNames(hookCodeFrame);
+
+                renderTemplateToken(t);
+
+                hookCodeFrame.addCode(currentCodeFrame.getCode());
+
+                // Switch back from hook CodeFrame to caller CodeFrame.
+                currentCodeFrame = callerCodeFrame;
+            }
+            case TemplateToken t -> {
+                // Use a nested CodeFrame.
+                CodeFrame callerCodeFrame = currentCodeFrame;
+                currentCodeFrame = CodeFrame.make(currentCodeFrame);
+
+                renderTemplateToken(t);
+
+                callerCodeFrame.addCode(currentCodeFrame.getCode());
+                currentCodeFrame = callerCodeFrame;
+            }
+            case AddNameToken(Name name) -> {
+                currentCodeFrame.addName(name);
+            }
+        }
+    }
+
+    private void renderTokenList(List<Token> tokens) {
+        CodeFrame codeFrame = currentCodeFrame;
+        for (Token t : tokens) {
+            renderToken(t);
+        }
+        if (codeFrame != currentCodeFrame) {
+            throw new RuntimeException("Internal error: CodeFrame mismatch.");
+        }
+    }
+
+    private void renderStringWithDollarAndHashtagReplacements(String s) {
```

Comment:
Hard to grasp the logic of that method. But I trust you on that :-) I leave it up to you if you want to improve readability to extract some of the logic to separate methods such that this method becomes easier to understand.

Reply:
I split out a part and added some more comments / examples.